sources/web/datalab/jupyter.ts

```diff
@@ -81,67 +88,93 @@ function pipeOutput(stream: NodeJS.ReadableStream, port: number, error: boolean)
-  server.childProcess.on('exit', exitHandler);
-  logging.getLogger().info('Jupyter process for user %s started with pid %d and args %j',
-                           userId, server.childProcess.pid, processArgs);
-
-  // Capture the output, so it can be piped for logging.
-  pipeOutput(server.childProcess.stdout, server.port, /* error */ false);
-  pipeOutput(server.childProcess.stderr, server.port, /* error */ true);
-
-  // Then create the proxy.
-  var proxyOptions: httpProxy.ProxyServerOptions = {
-    target: 'http://127.0.0.1:' + server.port
-  };
-
-  server.proxy = httpProxy.createProxyServer(proxyOptions);
-  server.proxy.on('proxyRes', responseHandler);
-  server.proxy.on('error', errorHandler);
-
-  return server;
+  tcp.waitUntilFree(port).then(
+    function() {
+      var userDir = userManager.getUserDir(userId);
+      if (!fs.existsSync(userDir)) {
+        fs.mkdirSync(userDir, parseInt('0755', 8));
+      }
+
+      var server: JupyterServer = {
+        userId: userId,
+        port: port,
+        notebooks: userDir,
+        portResolved: function() {},
+      };
+
+      function exitHandler(code: number, signal: string): void {
+        logging.getLogger().error('Jupyter process %d for user %s exited due to signal: %s',
+                                  server.childProcess.pid, userId, signal);
+        delete jupyterServers[server.userId];
+      }
+
+      var processArgs = appSettings.jupyterArgs.slice().concat([
+        '--port=' + server.port,
+        '--port-retries=0',
+        '--notebook-dir="' + server.notebooks + '"'
+      ]);
+
+      var notebookEnv: any = process.env;
+      if ('KG_URL' in notebookEnv && notebookEnv['KG_URL']) {
+        logging.getLogger().info(
+          'Found a kernel gateway URL of %s... configuring the notebook to use it', notebookEnv['KG_URL']);
+        processArgs = processArgs.concat([
+          '--NotebookApp.session_manager_class=nb2kg.managers.SessionManager',
+          '--NotebookApp.kernel_manager_class=nb2kg.managers.RemoteKernelManager',
+          '--NotebookApp.kernel_spec_manager_class=nb2kg.managers.RemoteKernelSpecManager'
+        ]);
+      }
+
+      var processOptions = {
+        detached: false,
+        env: notebookEnv
+      };
+
+      server.childProcess = childProcess.spawn('jupyter', processArgs, processOptions);
+      server.childProcess.on('exit', exitHandler);
+      logging.getLogger().info('Jupyter process for user %s started with pid %d and args %j',
+                               userId, server.childProcess.pid, processArgs);
+
+      server.portResolved = function() {
+        // Create the proxy.
+        var proxyOptions: httpProxy.ProxyServerOptions = {
+          target: 'http://127.0.0.1:' + port
+        };
+
+        server.proxy = httpProxy.createProxyServer(proxyOptions);
+        server.proxy.on('proxyRes', responseHandler);
+        server.proxy.on('error', errorHandler);
+
+        tcp.waitUntilUsed(server.port).then(
```

Comment:
The fact that the port was free before and after we spawned the jupyter process it is used should be safe enough. If you want to be even safer we can possibly call some libraries to make sure the port is used by the jupyter process we started.

The stdout message may not be reliable in the sense that a new jupyter version may choose to modify the message or removing the message and thus break us. It is not supposed to be used as part of the interface.